AVM 7 spec changes (#65)
algorandfoundation/specs: dev/TEAL.md

| `extract_uint16` | A uint16 formed from a range of big-endian bytes from A starting at B up to but not including B+2. If B+2 is larger than the array length, the program fails |
| `extract_uint32` | A uint32 formed from a range of big-endian bytes from A starting at B up to but not including B+4. If B+4 is larger than the array length, the program fails |
| `extract_uint64` | A uint64 formed from a range of big-endian bytes from A starting at B up to but not including B+8. If B+8 is larger than the array length, the program fails |
| `replace2 s` | Copy of A with the bytes starting at S replaced by the bytes of B. Fails if S+len(B) exceeds len(A) |
| `replace3` | Copy of A with the bytes starting at B replaced by the bytes of C. Fails if B+len(C) exceeds len(A) |

> | `base64_decode e` | Decode Base-64 encoded byte-array A. Fail if A is not properly encoded. |

Reviewer: ```suggestion
| `base64_decode e` | Decode Base-64 encoded byte-array A. Fail if A is not properly encoded. WARNING: Usage should be restricted to very rare use cases. |
```

Author: I'm very happy with this warning on `json_ref` but I think we can be a bit softer here, since at least b64 is well defined and not all that slow.  How about: "Users of `base64_decode` should consider whether their input could be supplied unencoded to save transaction space and computation within the AVM"

Reviewer: Indeed, "Users of base64_decode should consider whether their input could be supplied unencoded to save transaction space and computation within the AVM" looks good to me

> | `json_ref r` | Return key B's value from a [valid](jsonspec.md) utf-8 encoded json object string A. Fail if B's value can not be represented by the type R. |

Reviewer: ```suggestion
| `json_ref r` | Return key B's value from a [valid](jsonspec.md) utf-8 encoded json object string A. Fail if B's value can not be represented by the type R. WARNING: Usage should be restricted to very rare use cases. |
```

| 16 | TypeEnum | uint64 |      | See table below |
| 17 | XferAsset | uint64 |      | Asset ID |
| 18 | AssetAmount | uint64 |      | value in Asset's units |

> | 19 | AssetSender | []byte |      | 32 byte address. Moves asset from AssetSender if Sender is the Clawback address of the asset. |

Reviewer: ```suggestion
| 19 | AssetSender | []byte |      | 32 byte address. Used to move asset from AssetSender if Sender is the Clawback address of the asset. |
```

Author: Agree that it should be reworded to describe the field, rather than sounding like the field is a command.  How about:
"32 byte address. The source of assets if Sender is the Asset's Clawback address."

The change to "Asset's Clawback address" at the end is to match the syntax of "AssetAmount", which reads:
"value in Asset's units"

Maybe "value" that should be capitalized for AssetAmount.

Reviewer: "32 byte address. The source of assets if Sender is the Asset's Clawback address."

looks good

| 60 | CreatedAssetID | uint64 | v5  | Asset ID allocated by the creation of an ASA (only with `itxn` in v5). Application mode only |
| 61 | CreatedApplicationID | uint64 | v5  | ApplicationID allocated by the creation of an application (only with `itxn` in v5). Application mode only |
| 62 | LastLog | []byte | v6  | The last message emitted. Empty bytes if none were emitted. Application mode only |

> | 63 | StateProofPK | []byte | v6  | 64 byte state proof public key commitment |

Reviewer: Is it for key registration transaction
```suggestion
| 63 | StateProofPK | []byte | v6  | 64 byte state proof public key |
```
I'm afraid that the word "commitment" can be confusing. The fact that it is a commitment is mostly not relevant to the developer.

with the next instruction with, for example, `balance` and
`min_balance` checks. In v6, inner transactions may also perform
`keyreg` and `appl` effects. Inner `appl` calls fail if they attempt
to invoke a program with version less than v4, or if they attempt to

> opt-in to an app with a ClearState Program less than v4.

Reviewer: Is this change retroactive for v6?

Author: yes

For version 1, subsequent bytes after the varuint are program opcode bytes. Future versions could put other metadata following the version identifier.

It is important to prevent newly-introduced transaction types and
fields from breaking assumptions made by programs written before they
existed. If one of the transactions in a group will execute a program
whose version predates a transaction type or field that can violate
expectations, that transaction type or field must not be used anywhere

> in the transaction group. A v1 program included in a transaction group

Reviewer: ```suggestion
in the transaction group. 

Concretely, the above requirement is translated as follows:
A v1 program included in a transaction group
```

Author: I've committed that. If this seems good, please approve but hold off on merging, as per @algorandskiy

Reviewer 2: I withdraw my prev request to hold the merge - I've used the block size increase commit hash for v33 and any other final merge can be used for v34. Please go ahead and merge if it looks good.

that includes a ApplicationCall transaction, or a non-zero RekeyTo
field, will fail regardless of the other contents of the program
itself.

This requirement is enforced as follows:

* For every transaction, compute the earliest version that supports
  all the fields and values in this transaction.
  
* Compute the largest version number across all the transactions in a group (of size 1 or more), call it `maxVerNo`. If any transaction in this group has a program with a version smaller than `maxVerNo`, then that program will fail.

> In addition, applications must be v6 or greater to be eligible for

Reviewer: Wasn't this constraint removed?

Author: Correct, I missed this reference. I'll change.
